### PR TITLE
feat: add pre & post liquidation hooks

### PR DESCRIPTION
**Pull request by avniculae**
https://www.notion.so/voltz/Pre-Post-hooks-b08aeed29865472292c52096b83486f5?pvs=4

### CODE REVIEW

File: core/src/libraries/LiquidationBidPriorityQueue.sol [hunk 1]
```diff
@@ -17,6 +17,7 @@ library LiquidationBidPriorityQueue {
 
     struct LiquidationBid {
         uint128 liquidatorAccountId;
+        address hookAddress;
```
Comment (0xZenus) [inline core/src/libraries/LiquidationBidPriorityQueue.sol:20]: Why is the hook address defined per bid?

File: core/src/modules/LiquidationModule.sol [hunk 2]
```diff
@@ -70,6 +71,13 @@ contract LiquidationModule is ILiquidationModule {
         // grab the liquidatable account and check its existance
         Account.Data storage account = Account.exists(liquidatableAccountId);
 
+        if (liquidationBid.hookAddress != address(0)) {
```
Comment (avniculae) [inline core/src/modules/LiquidationModule.sol:74]: add todo

File: core/src/libraries/LiquidationBidPriorityQueue.sol (hunk 1 above)
Comment (arturbeg) [reply core/src/libraries/LiquidationBidPriorityQueue.sol:20]: each liquidator may want to use custom hook implementation, e.g. one may want to do a flash loan, liquidate, withdraw and then return the flash loan, do you think there's an alternative way of approaching this?

File: core/src/interfaces/external/ILiquidationHook.sol [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+You may obtain a copy of the License at
+
+https://github.com/Voltz-Protocol/v2-core/blob/main/core/LICENSE
+*/
+pragma solidity >=0.8.19;
+
+import {LiquidationBidPriorityQueue} from "../../libraries/LiquidationBidPriorityQueue.sol";
+import {IERC165} from "@voltz-protocol/util-contracts/src/interfaces/IERC165.sol";
+
+interface ILiquidationHook is IERC165 {
```
Comment (arturbeg) [inline core/src/interfaces/external/ILiquidationHook.sol:13]: additionally the hook will need to be granted "ADMIN" access by the account to be able to withdraw and deposit on its behalf? worth adding it in natspec?

File: core/src/modules/LiquidationModule.sol [hunk 4]
```diff
@@ -70,6 +81,17 @@ contract LiquidationModule is ILiquidationModule {
...
+            if (
+                ILiquidationHook(liquidationBid.hookAddress).preLiquidationHook(
+                    liquidatableAccountId,
+                    liquidationBid
+                ) != ILiquidationHook.preLiquidationHook.selector
+            ) {
+                revert InvalidPreLiquidationHookResponse();
```
Comment (arturbeg) [inline core/src/modules/LiquidationModule.sol:91]: what's this check trying to achieve, can it end up blocking the queue pop?

File: core/src/interfaces/external/ILiquidationHook.sol (hunk 3 above)
Comment (avniculae) [reply core/src/interfaces/external/ILiquidationHook.sol:13]: yes, that's correct.

File: core/src/modules/LiquidationModule.sol (hunk 4 above)
Comment (avniculae) [reply core/src/modules/LiquidationModule.sol:91]: This is inside the `executeLiquidationBid` function, which is called using low-level `.call()` inside the `executeTopRankedLiquidationBid`. Therefore,  popping from the queue will not be blocked if this reverts.

Comment (avniculae) [reply core/src/modules/LiquidationModule.sol:91]: To answer the first question though, this is an extra check inspired from Uniswap v4 (see [this](https://github.com/Uniswap/v4-core/blob/0095e0848098c3e32e016eac6d2537b67aa47358/contracts/PoolManager.sol#L115))

File: core/src/interfaces/external/ILiquidationHook.sol (hunk 3 above)
Comment (avniculae) [reply core/src/interfaces/external/ILiquidationHook.sol:19]: [b4ea582](https://github.com/Voltz-Protocol/v2-core/pull/171/commits/b4ea582b8a03d7a9bece5ca2d59e5a8f508aeb82)

File: core/src/modules/LiquidationModule.sol (hunk 4 above)
Comment (arturbeg) [reply core/src/modules/LiquidationModule.sol:91]: got it, thank you

**Review by arturbeg — APPROVED** [review]

**Pull request merged**